### PR TITLE
[asl] Typing v0.2

### PR DESCRIPTION
**Pull request by HadrienRenaud**
This PR brings:

- [x] An implementation of the type algebra and relations in `asllib/Types.ml`
- [x] Use of this type-algebra for type-checking in `asllib/Typing.ml`
- [x] Implementation of an expression reduction algorithm in `asllib/StaticInterpreter.ml`
- [x] Use of a shared environment type defined in `asllib/env.ml`
- [x] Enforce far more type-checking rules, such as immutable storage elements.

### CODE REVIEW

**Comment by maranget** [comment]
Hi @HadrienRenaud. When I was talking about scopes, I was mostly thinking about block structures. That is, variables defined in a block are no longer defined after block end. Additionally, it seems that `stmt_list` defines such a block.

I cannot explain it better than with commit   8b146b3, an implementation attempt of such a block definition in the type checker. BTW I am wondering whether type declarations should follow the same scoping rules or not.

**Comment by HadrienRenaud** [comment]
Oh I understand.

It might be counter-productive to do this while this PR is not merged though, because changes would need to be rebased. I suggest we wait until this PR is closed? (I might even do it in the mean-time)

File: asllib/Typing.ml [hunk 1]
```diff
@@ -1351,24 +1334,33 @@ module Annotate (C : ANNOTATE_CONFIG) = struct
...
+    (*
+        See rule JFRD:
+           A local identifier declared with var, let or constant is in scope
+           from the point immediately after its declaration until the end of the
+           immediately enclosing block.
+
+        From that follows that we can discard the environment at the end of an
+        enclosing block.
+    *)
+    best_effort s (fun _ -> annotate_stmt env return_type s |> fst)
```
[the remark this answers is not in the document]
Comment (maranget) [reply asllib/Typing.ml:1356]: I think it does, I'll try tomorrow

**Comment by HadrienRenaud** [comment]
Hi @maranget 

Could you have a look at this PR? Specifically, if you could give me feedback on:
 - `asllib/Types.ml` which should implement the types algebra
 - the typing of loops in `asllib/Typing.ml`
 - `asllib/StaticInterpreter.ml` for the expression reduction code (I'm not very happy with this code, I can completely rework on that following your feedback)

There are still a lot of `TODO` comments, but none of them are quickly fixable, so I'm ok with letting them live until further developments.

Things that should be doable from this point:
 - [ ] Global variables
 - [ ] Subtype declarations
 - [ ] Exceptions
 - [ ] Bit-vector masks
 - [ ] Pure functions
 - [ ] subprogram types checking

**Comment by HadrienRenaud** [comment]
`herd7` by default has a very permissive asl type-checker that makes type-checking difficult. This is to allow the aarch64 shared-pseudocode to keep undefined things on non-taken control-flow pathes.

On the contrary, `aslseq` has by default the full type-checking options (see `aslseq -h`). It outputs as expected:

```
❯ dune exec aslseq -- test-luc.asl 
File test-luc.asl, line 6, characters 9 to 10:
ASL Error: Undefined identifier: 's'
```

(`test-luc.asl` is the litmus test you shared without its litmus header, line 6 is `return s;`).

Maybe it would be interesting to have a herd7 variant to add type-checking?
Or to only disable strict type-checking if the option `-variant AArch64+ASL` is on?

**Comment by jalglave** [comment]
hi @Renaud, ***@***.***> we need herd to support the same features as asl-seq, e.g. type checking.
Thanks,
Jade
________________________________

**Comment by HadrienRenaud** [comment]
> we need herd to support the same features as asl-seq, e.g. type checking.

This is already implemented, my sentence was not clear, sorry about that. We just need to pass the right flags to the interpreter to make type-checking strict by default on `herd7`. My question was more on the right flags to do that: do we want more `herd7` variant? Or do we just assume with current variants?


PS: there still is a need for permissive type-checking because of the ASL backend of the AArch64 architecture, which as of now, does not type-checks in herd7 mainly because I did not implement all the architecture defined functions and constants. This would require quite a big amount of work.

**Comment by HadrienRenaud** [comment]
Hi @maranget 

Nice work on those commits. However I think we should stop committing on this branch and try to get this merged. What do you think?

**Comment by maranget** [comment]
Probably you are right....

**Comment by maranget** [comment]
Hi, I am cleaning may last commit.  Then, do you think the PR is ready?

**Comment by HadrienRenaud** [comment]
I think so! I didn't check that tests pass and that it compiles on 4.08 though.

Can I let you squash at your convenience and then merge? Thanks :)

**Comment by maranget** [comment]
Hi @HadrienRenaud. The PR is now squashed and tested for 4.08.0 compatibility. Additionally `make test` is fine.

**However**, ` make test-aarch64-asl` fails. Which is a pity as this test is fine on the master branch. As this test is optional   test,  we can merge without severe damage. But, I'd rather have this problem fixed, if it is easily fixable,,,

**Comment by HadrienRenaud** [comment]
Thanks @maranget! I'll have a look

**Comment by HadrienRenaud** [comment]
Hi @maranget 

I've fixed `make test-aarch64-asl`, it was not really "easily" but it is done.
Should we merged?

**Comment by maranget** [comment]
That's fine, thanks a lot.

**Pull request merged**